README.md
<p align="center">
  <img src="https://github.com/tbrus/install-ifcopenshell-python/blob/master/logo.png?raw=true"/>
</p>

# install-ifcopenshell-python

![PyPI - License](https://img.shields.io/pypi/l/install-ifcopenshell-python)
![PyPI](https://img.shields.io/pypi/v/install-ifcopenshell-python)

[![pre-commit](https://img.shields.io/badge/pre--commit-enabled-brightgreen?logo=pre-commit&logoColor=white)](https://github.com/pre-commit/pre-commit)
[![Code style](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)
[![Imports: isort](https://img.shields.io/badge/%20imports-isort-%231674b1?style=flat&labelColor=ef8336)](https://pycqa.github.io/isort/)

This package is a "part" of another repo: 
[pyifc](https://github.com/tbrus/pyifc/).
 
[ifcopenshell on github](https://github.com/IfcOpenShell/IfcOpenShell)  
[ifcopenshell website](http://ifcopenshell.org/python)

Unfortunately **ifcopenshell** is not distributed via PyPi. Hence you can get 
error while installing **pyifc**:

```bash
ModuleNotFoundError: No module named 'ifcopenshell'
```

To address this issue you can use 
[install-ifcopenshell-python](https://github.com/tbrus/install-ifcopenshell-python).

## Table of Contents

[Installation](https://github.com/tbrus/install-ifcopenshell-python#installation) |
[Usage](https://github.com/tbrus/install-ifcopenshell-python#usage) |
[References](https://github.com/tbrus/install-ifcopenshell-python#license)

## Installation

Use the package manager [pip](https://pip.pypa.io/en/stable/) to install
**install-ifcopenshell-python**.

```bash
pip install install-ifcopenshell-python
```

## Usage

Ifcopenshell is downloaded via hard coded URL (v0.6.0).

No arguments are needed. System, architecture and python version are determined 
automatically. You can simply run:

```bash
python -m install_ifcopenshell_python
```
Notice that here underscores are used, not hyphens.

However, if you would like to, you can determine your own arguments. You can 
run:

```bash
python -m install_ifcopenshell_python -h
```

To produce output:

```bash
usage: install_ifcopenshell_python [-h] [-s {linux,win,macos}] [-a {32,64}] [-v {37,38,39}]

Download ifcopenshell based on platform system, platform architecture and running python version.

optional arguments:
  -h, --help            show this help message and exit
  -s {linux,win,macos}, --system {linux,win,macos}
                        Platform system.
  -a {32,64}, --architecture {32,64}
                        Platform architecture.
  -v {37,38,39}, --python-version {37,38,39}
                        Running python version; string of major and minor version, e.g. '39'. pyifc supports python >=
                        3.7.
```

## References

[1] [Logo](https://github.com/tbrus/install-ifcopenshell-python/blob/master/logo.png) 
is created based on logo of [pyifc](https://github.com/tbrus/pyifc/) 
and logo of [ifcopenshell](http://ifcopenshell.org/python).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.0] - 2021-12-28

Installation is done via hard coded URL from aws; ifcopenshell (v0.6.0):  
```python
url = https://s3.amazonaws.com/ifcopenshell-builds/ifcopenshell-python-{args.python_version}-v0.6.0-517b819-{args.system}{args.architecture}.zip
```


install_ifcopenshell_python/__main__.py
"""
install_ifcopenshell_python.__main__
------------------------------------

Install ifcopenshell for python based on hard coded aws URL.

NOTE: ifcopenshell will not be visible via command `pip list`.
"""

import argparse
import os
import platform
import shutil
import site
import sys
import urllib.request
import zipfile


class SystemNotFoundError(Exception):
    pass


class PythonVersionError(Exception):
    pass


def get_system():
    """Get current platform system.

    Raises:
        SystemNotFoundError: if `system` is not any of the values:
            'linux', 'windows', 'darwin'.

    Returns:
        str: platform system.
    """
    if platform.system().lower() == "linux":
        system = "linux"
    elif platform.system().lower() == "windows":
        system = "win"
    elif platform.system().lower() == "darwin":
        system = "macos"
    else:
        msg = (
            f"{platform.system().lower()} is not the valid system name. "
            f"You can specify system manually via argument in command line."
        )
        raise SystemNotFoundError(msg)
    return system


def get_architecture():
    """Get platform architecture.

    Returns:
        str: platform architecture. One of values: '32' or '64'.
    """
    architecture_bit = platform.architecture()[0]
    architecture_list = [i for i in architecture_bit if i.isdigit()]
    architecture = "".join(architecture_list)
    return architecture


def get_python_version():
    """Get running python version.

    This function returns string of 2 digits. The first one determines
    the major version and the second one determines the minor version of
    running python, e.g. '38', '39'.
    pyifc supports python >= 3.7.

    Returns:
        str: current python version
    """
    python_major = str(sys.version_info.major)
    python_minor = str(sys.version_info.minor)
    python_version = python_major + python_minor
    if int(python_version) < 37:
        msg = (
            f"python{python_major}.{python_minor} is not supported "
            f"by pyifc. Install python >= 3.7) on your computer."
        )
        raise PythonVersionError(msg)
    return python_version


def get_args():
    """Get values for system, architecture and python version
    specified by the user via command line. Default values are
    determined automatically.

    Possible arguments typed via command line are:
        -s --system -  platform system. Possible values: 'linux',
            'win', 'macos'.
        -a --architecture - platform architecture. Possible values:
              '32' and '64'.
        -v --python_version - version of running python. This should be
            a string of two digits: the major and minor versions of
            running python. Possible values: '37', '38', '39' (pyifc
            supports python >= 3.7).

    Returns:
        [argparse.Namespace]: Namespace containing 3 values:
            - str - args.system
            - str - args.architecture
            - str - args.python_version
    """
    description = (
        "Download ifcopenshell based on platform system, "
        "platform architecture and running python version."
    )
    parser = argparse.ArgumentParser(description=description)
    parser.add_argument(
        "-s",
        "--system",
        help="Platform system.",
        type=str,
        choices=["linux", "win", "macos"],
        default=get_system(),
    )
    parser.add_argument(
        "-a",
        "--architecture",
        help="Platform architecture.",
        type=str,
        choices=["32", "64"],
        default=get_architecture(),
    )
    parser.add_argument(
        "-v",
        "--python-version",
        help=(
            "Running python version; string of major and minor version, "
            "e.g. '39'. pyifc supports python >= 3.7."
        ),
        type=str,
        choices=["37", "38", "39"],
        default=get_python_version(),
    )
    args = parser.parse_args()
    return args


def main():
    """The main function of this script.

    Downloads ifcopenshell.zip, extracts and removes zip file, moves dir
    with ifcopenshell package to site-packages directory of currently
    running venv. It forces moving (overwrites package if exists).
    """
    # Get url for given system, architecture and python version
    args = get_args()
    print("Getting arguments...")
    print(
        f"Arguments: system='{args.system}', "
        f"architecture='{args.architecture}', "
        f"python_version='{args.python_version}'"
    )
    url = f"https://s3.amazonaws.com/ifcopenshell-builds/ifcopenshell-python-{args.python_version}-v0.6.0-517b819-{args.system}{args.architecture}.zip"

    # Check whether url is valid
    print(f"Checking if url is valid...")
    print(url)
    try:
        response = urllib.request.urlopen(url)
        code = response.getcode()
        print(f"Status: {code}")
        print("Given url is valid")
    except Exception as e:
        print("Given url is not valid")
        print(f"Error:  {url} : {str(e)}")

    # Download ifcopenshell package, unzip and delete zip file
    zip_filename = "ifcopenshell.zip"
    print("Downloading ifcopenshell...")
    urllib.request.urlretrieve(url, zip_filename)
    print("Extracting files...")
    with zipfile.ZipFile("ifcopenshell.zip", "r") as zip_ref:
        zip_ref.extractall()
    print("Removing irrelevant files...")
    if os.path.exists(zip_filename):
        os.remove(zip_filename)

    # Move ifcopenshell package directory to site-packages directory
    # of currently running venv
    # If package exists in target directory - remove it and move newly
    # downloaded package
    print("Moving files...")
    source_dir = os.path.join(os.getcwd(), "ifcopenshell")
    if args.system == 'win':
        target_dir = site.getsitepackages()[1]
    else:
        target_dir = site.getsitepackages()[0]
    target = os.path.join(target_dir, "ifcopenshell")
    if os.path.exists(target):
        print("Found ifcopenshell in packages")
        print("Removing found files...")
        shutil.rmtree(target)
        print("Successfully removed files")
        print("Moving newly downloaded files...")
    shutil.move(source_dir, target_dir)
    print("Ifcopenshell successfully moved")
    print(
        "\nWARNING: ifcopenshell is now available via import, but ",
        "not via pip list",
    )


if __name__ == "__main__":
    main()